Refactor multiplayer background to remove selected item bindable

smoogipoo opened the pull request:
Split off from https://github.com/ppy/osu/pull/32250

As I'm splitting this PR out, I'm going to treat `RoomSubScreen` as if it's already `MultiplayerMatchSubScreen`, as that is its only remaining inheritor.

smoogipoo committed:
Refactor multiplayer background to remove selected item bindable

smoogipoo committed:
Make class partial

bdach commented:
You know, the code here is fine and all, no objections, just one simple question though - *where on the room sunscreen **is** the map background*, exactly?

I'm pretty sure it's not there. And it's also not there on master.

I briefly touched on this previously in https://github.com/ppy/osu/pull/32458#issuecomment-2735957740, but therein it's a secondary concern because while the background *is* also missing on the panel, the panel also displays other information that does display correctly. But on a PR that touches the *background specifically* it's a bit difficult to go "yep all good" without question *when the background is a static image but there is things being done that attempt to pretend it's not*.

Is this going to be addressed at some point?

smoogipoo commented:
That's... a good question. I'm going to fix that here since it changes how this code is structured a bit.

smoogipoo committed:
Fix room background not working in multiplayer

At the core of the problem is that the multiplayer server does not
serialise beatmap covers to clients -- only the beatmap ids. Because of
this, any components that need to display the background should query it
from an online source first (i.e. via `BeatmapLookupCache`).

There is a slightly tricky situation here formed of two parts, which
I'll try to explain below.

`Background.Sprite` is exposed publicly and some inheritors override the
sprite's texture in a similar fashion to the way this changeset does.
While I frankly believe this is unnaceptable from an encapsulation point
of view, I've gone for consistency in this regard.

The other fail case is `UpdateableBeatmapBackgroundSprite`. Contrary to
its name, that object is _not_ a `Sprite` - it is a
`ModelBackedDrawable` that _contains_ a `Sprite`. The logic in this PR
could be extracted into a separate object similar to
`OnlineBeatmapSetCover` (an actual `Sprite`), but even so it would
require `Background` to provide a path for overriding its contained
`Sprite`.
I went through the path above originally with the changes visible in
https://github.com/smoogipoo/osu/tree/fix-mp-backgrounds, but it looks
quite daunting and I feel like there would be a lot of emphasis on the
correct abstraction model for `Background`, of which I'm not entirely
sure of.

Instead of dealing with both of the above, this commit presents a
direct; local resolution to the problem.

smoogipoo commented on osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs:
```diff
@@ -16,65 +21,57 @@ namespace osu.Game.Screens.OnlinePlay.Components
+            cancellationSource?.Cancel();
+            cancellationSource = new CancellationTokenSource();
 
-                string? lastCover = (background?.Beatmap?.BeatmapSet as IBeatmapSetOnlineInfo)?.Covers.Cover;
-                string? newCover = (beatmap?.BeatmapSet as IBeatmapSetOnlineInfo)?.Covers.Cover;
```
This now dedupes on beatmap id (inside `PlaylistItem` setter) rather than cover resource. It does mean that the background will be reloaded if you select multiple beatmaps from the same set, but I don't think this is particularly important to consider - multi/playlists generally don't have multiple beatmaps from the same set in sequence.

smoogipoo commented on osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs:
```diff
@@ -16,65 +21,57 @@ namespace osu.Game.Screens.OnlinePlay.Components
-        private void updateBackground()
+        private void loadNewBackground()
         {
-            Schedule(() =>
```
Schedule removed, doesn't look required.

bdach replied:
Seems to have originated in a distant past wherein screens would just arbitrarily bind to parts of the room (https://github.com/ppy/osu/commit/44e157447c7d005079e2ede801ee6ad6fe5a802c), so probably fine to remove

smoogipoo commented on osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs:
```diff
@@ -16,65 +21,57 @@ namespace osu.Game.Screens.OnlinePlay.Components
 
-                if (LoadState > LoadState.Ready)
-                    updateBackground();
+                if (LoadState >= LoadState.Ready)
```
Small difference here - there are no cases of `LoadState > LoadState.Ready` in the game. All others use `>=`. In other words, original code was slightly incorrect.

Though in practice this is unlikely to be hit.

smoogipoo committed:
Make class partial

bdach commented on osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs:
```diff
@@ -89,5 +86,47 @@ public override bool OnExiting(ScreenExitEvent e)
+                    if (coverImage != null)
+                        Sprite.Texture = textures.Get(coverImage);
+                }
+                catch
+                {
+                }
```
would probably log something at least

bdach commented on osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs:
```diff
@@ -89,5 +86,47 @@ public override bool OnExiting(ScreenExitEvent e)
+                Sprite.Texture = beatmapManager.DefaultBeatmap.GetBackground();
+            }
+
+            public override bool Equals(Background? other)
+                => other is DefaultBackground;
```
What's this doing?

- The base implementation of this equality member seems to be doing the exact same thing as this override
- To the best of my knowledge, `IEquatable<Background>` is not really used outside of `BackgroundScreenDefault`

smoogipoo replied:
I knew this was gonna be commented on. It was added because this used to be a `public` class, but unused in this case - will remove.

bdach replied:
Yeah sorry if this is seen as me being a stickler but equality members are similar to `.ToString()` overrides in that once they're in nobody wants to ever touch them again lest there be some stupid breakage somewhere resulting from the removal that you can't reasonably check for because how do you even, so I prefer to crack down on these as soon as possible.

bdach commented on osu.Game/Screens/OnlinePlay/Components/OnlinePlayBackgroundScreen.cs:
```diff
@@ -89,5 +86,47 @@ public override bool OnExiting(ScreenExitEvent e)
+            {
+                try
+                {
+                    APIBeatmap? beatmap = lookupCache.GetBeatmapAsync(beatmapId, cancellationToken).GetResultSafely();
```
This is a vague and mostly unsubstantiated worry so no action required, but I do kinda worry if this is gonna hold up. I don't think we've done hard async-to-sync `.GetResultSafely()` calls in `[LongRunningLoad]` BDLs anywhere. Especially when API calls (hidden beneath task completion sources) are involved as well.

It seems that in practice this will be a cache hit anyway so probably not that relevant but I guess we'll see how well this does...

smoogipoo replied:
When you say "hold up", in what way do you mean/from what angle? I ask because it's possible there's something I haven't considered.

We don't really _need_ to use `GetResultSafely()`. The primary thinking here was `OnlineBeatmapSetCover` is also a LRL and the "long" part is how `OnlineStore` behaves internally down this sync path. Given that this is also an online request of sorts, it sounded like it should behave similarly.

> in practice this will be a cache hit anyway so

I wouldn't rely on this though. Hard to say whether the behaviour will be identical in creating vs joining rooms, for example.

smoogipoo committed:
Apply suggestions from review
